### PR TITLE
Event Tracking:  For statement

### PR DESCRIPTION
**Pull request by zhang1002**
Generate KyuubiStatementInfo. This object includes the following elements:
1. statement
2. statementId
3. appId
4. sessionId
5. exception: this statement's exception
6. physicalPlan
7. stateTime: contains each state and the time occurrence
8. sparkUser

Those data was packaged in KyuubiStatementInfo and we store those object in mem.
You can get some summary data from this object.

### CODE REVIEW

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala [hunk 1]
```diff
@@ -58,6 +60,18 @@ class ExecuteStatement(
 
   private val operationListener: SQLOperationListener = new SQLOperationListener(this, spark)
 
+  var kStatement: KStatement = new KStatement(
+    statement, getHandle.identifier.toString,
+    spark.sparkContext.applicationId,
+    session.getTypeInfo.identifier.toString,
+    OperationState.INITIALIZED.toString,
+    new Date().getTime)
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala:68]: System.currentTimeMillis?

File: externals/kyuubi-spark-monitor/src/main/scala/org.apache.spark.kyuubi/entity/KyuubiStatementInfo.scala [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+case class KyuubiStatementInfo(
+    statementId: String,
+    statement: String,
+    appId: String,
+    sessionId: String,
+    var physicalPlan: String,
+    sparkUser: String,
+    var exception: String,
+    var stateTimeMap: Map[String, Long])
```
Comment (ulysses-you) [inline externals/kyuubi-spark-monitor/src/main/scala/org.apache.spark.kyuubi/entity/KyuubiStatementInfo.scala:30]: if this field can be modified, we can add into class.
```
case class {
  var xxx: Option[String];
}
```

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala [hunk 3]
```diff
@@ -58,6 +61,12 @@ class ExecuteStatement(
 
   private val operationListener: SQLOperationListener = new SQLOperationListener(this, spark)
 
+  var kyuubiStatementInfo = KyuubiStatementInfo(
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala:64]: `var` -> `private val`

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala [hunk 4]
```diff
@@ -156,6 +167,17 @@ class ExecuteStatement(
...
+
+  override def setState(newState: OperationState): Unit = {
+    super.setState(newState)
+    KyuubiStatementMonitor.addEachStateTimeForOperationid(
+      statementId, newState.toString, lastAccessTime)
+  }
+
+  override def setOperationException(opEx: KyuubiSQLException): Unit = {
+    KyuubiStatementMonitor.addOperationExceptionByOperationId(statementId, opEx.toString)
+    super.setOperationException(opEx)
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala:179]: please keep the same order.
```
super.setOperationException(opEx)
KyuubiStatementMonitor.addOperationExceptionByOperationId(statementId, opEx.toString)
```

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala [hunk 5]
```diff
@@ -58,6 +61,12 @@ class ExecuteStatement(
 
   private val operationListener: SQLOperationListener = new SQLOperationListener(this, spark)
 
+  var kyuubiStatementInfo = KyuubiStatementInfo(
+    statementId, statement, spark.sparkContext.applicationId,
+    session.getTypeInfo.identifier.toString, null,
+    spark.sparkContext.sparkUser, null, Map(state.toString->lastAccessTime))
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala:67]: `Map(state.toString->lastAccessTime)` -> `Map(state.toString -> lastAccessTime)`

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala [hunk 6]
```diff
@@ -19,7 +19,10 @@ package org.apache.kyuubi.engine.spark.operation
 
 import java.util.concurrent.{RejectedExecutionException, ScheduledExecutorService, TimeUnit}
 
-import org.apache.spark.kyuubi.SQLOperationListener
+import scala.collection.mutable.Map
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala:22]: I think we don't need mutable Map here

File: externals/kyuubi-spark-monitor/src/main/scala/org.apache.spark.kyuubi/entity/KyuubiStatementInfo.scala (hunk 2 above)
Comment (ulysses-you) [reply externals/kyuubi-spark-monitor/src/main/scala/org.apache.spark.kyuubi/entity/KyuubiStatementInfo.scala:30]: and same with `stateTimeMap`, I'd like move it into class

File: externals/kyuubi-spark-monitor/src/main/scala/org.apache.spark.kyuubi/KyuubiStatementMonitor.scala [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+object KyuubiStatementMonitor {
+
+  private val operationStatementMap = new ConcurrentHashMap[String, KyuubiStatementInfo]()
+
+//  private var endStateList = new util.ArrayList[String]()
+//  endStateList.add(OperationState.FINISHED.toString)
+//  endStateList.add(OperationState.CANCELED.toString)
+//  endStateList.add(OperationState.CLOSED.toString)
+//  endStateList.add(OperationState.ERROR.toString)
+//  endStateList.add(OperationState.TIMEOUT.toString)
```
Comment (ulysses-you) [inline externals/kyuubi-spark-monitor/src/main/scala/org.apache.spark.kyuubi/KyuubiStatementMonitor.scala:33]: please remove unused code

File: externals/kyuubi-spark-monitor/src/main/scala/org.apache.spark.kyuubi/KyuubiStatementMonitor.scala [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+  def addStatementDetailForOperationId(
+      operationId: String, kyuubiStatementInfo: KyuubiStatementInfo): Unit = {
+    operationStatementMap.putIfAbsent(operationId, kyuubiStatementInfo)
+  }
+
+  def addEachStateTimeForOperationid(operationId: String, state: String, time: Long): Unit = {
+    operationStatementMap.get(operationId).stateTimeMap.put(state, time)
+    print(1231)
```
Comment (ulysses-you) [inline externals/kyuubi-spark-monitor/src/main/scala/org.apache.spark.kyuubi/KyuubiStatementMonitor.scala:42]: please remove unused code

File: externals/kyuubi-spark-monitor/src/main/scala/org.apache.spark.kyuubi/KyuubiStatementMonitor.scala [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.kyuubi
+
+import java.util.concurrent.ConcurrentHashMap
+
+import org.apache.spark.kyuubi.entity.KyuubiStatementInfo
+
+object KyuubiStatementMonitor {
```
Comment (ulysses-you) [inline externals/kyuubi-spark-monitor/src/main/scala/org.apache.spark.kyuubi/KyuubiStatementMonitor.scala:24]: it's better to add some comment at class description.

File: externals/kyuubi-spark-monitor/src/main/scala/org.apache.spark.kyuubi/entity/KyuubiStatementInfo.scala [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+ * @param sparkUser
+ * @param exception
+ * @param stateTimeMap: store this statement's every state and the time of occurrence
+ */
+case class KyuubiStatementInfo(
+    statementId: String,
+    statement: String,
+    appId: String,
+    sessionId: String,
+    var physicalPlan: String,
```
Comment (yaooqinn) [inline externals/kyuubi-spark-monitor/src/main/scala/org.apache.spark.kyuubi/entity/KyuubiStatementInfo.scala:39]: I think we can track `queryExecution` directly in memory

File: externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/KyuubiStatementMonitor.scala [hunk 11]
```diff
@@ -15,11 +15,11 @@
  * limitations under the License.
  */
 
-package org.apache.spark.kyuubi
+package org.apache.kyuubi
 
 import java.util.concurrent.ConcurrentHashMap
 
-import org.apache.spark.kyuubi.entity.KyuubiStatementInfo
+import org.apache.kyuubi.entity.KyuubiStatementInfo
```
Comment (yaooqinn) [inline externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/KyuubiStatementMonitor.scala:22]: org.apache.kyuubi.engine.spark.monitor.KyuubiStatementInfo

**Comment by yaooqinn** [comment]
How about add some unit tests?

File: externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/entity/KyuubiStatementInfo.scala [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+ * @param sparkUser
+ * @param exception
+ * @param stateTimeMap: store this statement's every state and the time of occurrence
+ */
+case class KyuubiStatementInfo(
+    statementId: String,
+    statement: String,
+    appId: String,
+    sessionId: String,
+    var physicalPlan: QueryExecution,
```
Comment (yaooqinn) [inline externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/entity/KyuubiStatementInfo.scala:45]: var queryExecution: QueryExecutio

File: externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/entity/KyuubiStatementInfo.scala [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    sparkUser: String,
```
Comment (yaooqinn) [inline externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/entity/KyuubiStatementInfo.scala:46]: remove this, we can manage this in session info

File: externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/entity/KyuubiStatementInfo.scala [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    var exception: KyuubiSQLException,
+    stateTimeMap: Map[OperationState, Long])
```
Comment (yaooqinn) [inline externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/entity/KyuubiStatementInfo.scala:48]: stateToTime

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala [hunk 15]
```diff
@@ -156,6 +167,16 @@ class ExecuteStatement(
     spark.sparkContext.removeSparkListener(operationListener)
     super.cleanup(targetState)
   }
+
+  override def setState(newState: OperationState): Unit = {
+    super.setState(newState)
+     kyuubiStatementInfo.stateTimeMap.put(newState, lastAccessTime)
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala:173]: remove ` `

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala [hunk 16]
```diff
@@ -58,6 +62,12 @@ class ExecuteStatement(
 
   private val operationListener: SQLOperationListener = new SQLOperationListener(this, spark)
 
+  private val kyuubiStatementInfo = KyuubiStatementInfo(
+    statementId, statement, spark.sparkContext.applicationId,
+    session.getTypeInfo.identifier.toString, null,
+    spark.sparkContext.sparkUser, null, Map(state -> lastAccessTime))
+  KyuubiStatementMonitor.addStatementDetailForOperationId(statementId, kyuubiStatementInfo)
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala:69]: could this lead to memory leak?

File: externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/entity/KyuubiStatementInfo.scala [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+ * @param physicalPlan: contains physicalPlan, logicPlan and so on
+ * @param sparkUser
+ * @param exception
+ * @param stateTimeMap: store this statement's every state and the time of occurrence
+ */
+case class KyuubiStatementInfo(
+    statementId: String,
+    statement: String,
+    appId: String,
+    sessionId: String,
```
Comment (yaooqinn) [inline externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/entity/KyuubiStatementInfo.scala:44]: use session handle?

**Review by yaooqinn — DISMISSED** [review]

File: externals/kyuubi-spark-monitor/src/test/resources/test-scheduler-pool.xml [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+        <minShare>2</minShare>
+        <weight>1</weight>
+        <schedulingMode>FAIR</schedulingMode>
+    </pool>
+    <pool name="p1">
+        <minShare>0</minShare>
+        <weight>1</weight>
+        <schedulingMode>FAIR</schedulingMode>
+    </pool>
+</allocations>
```
Comment (ulysses-you) [inline externals/kyuubi-spark-monitor/src/test/resources/test-scheduler-pool.xml:30]: why we need this file ?

File: externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/KyuubiStatementMonitor.scala [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+
+package org.apache.kyuubi.engine.spark.monitor
+
+import java.util.concurrent.ArrayBlockingQueue
+
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.engine.spark.monitor.entity.KyuubiStatementInfo
+
+// TODO: Thread Safe need to consider
+object KyuubiStatementMonitor extends Logging{
```
Comment (ulysses-you) [inline externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/KyuubiStatementMonitor.scala:26]: nit: `Logging{` -> `Logging {`

File: externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/KyuubiStatementMonitor.scala [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+
+  /**
+   * This function is used for putting kyuubiStatementInfo into blockingQueue(statementQueue).
+   * Every time we put an item into this queue, we should judge this queue's current size at first.
+   * If the size is less than threshold, we need to remove items from this queue.
+   * @param kyuubiStatementInfo
+   */
+  // TODO: Lack size type threshold and time type threshold
+  def putStatementInfoIntoQueue(kyuubiStatementInfo: KyuubiStatementInfo): Unit = {
+    if (kyuubiStatementQueue.size() >= 7) {
```
Comment (ulysses-you) [inline externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/KyuubiStatementMonitor.scala:49]: I'd like to define a constant:
```
val STATEMEMT_CAPACITY = 10
val FORCE_DUMP_STATEMENT_THREDHOLD = 7
...
```

File: externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/KyuubiStatementMonitor.scala [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+    info(s"Add kyuubiStatementInfo into queue is [$isSuccess], " +
+      s"statementId is [${kyuubiStatementInfo.statementId}]")
+  }
+
+  /**
+   * This function is used for removing kyuubiStatementInfo from blockingQueue(statementQueue)
+   * and dumpping them to a file by threshold.
+   */
+  // TODO: Need ensure those items have finished. If not, we should put them into this queue again.
+  def removeAndDumpStatementInfoFromQueue(): Unit = {
```
Comment (ulysses-you) [inline externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/KyuubiStatementMonitor.scala:62]: `private[kyuubi]`

File: externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/entity/KyuubiStatementInfo.scala [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+ * @param statement
+ * @param appId
+ * @param sessionId
+ * @param queryExecution: contains physicalPlan, logicPlan and so on
+ * @param exception
+ * @param stateToTime: store this statement's every state and the time of occurrence
+ */
+case class KyuubiStatementInfo(
+    statementId: String,
+    statement: String,
```
Comment (ulysses-you) [inline externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/entity/KyuubiStatementInfo.scala:41]: let's say: `query: String`

File: externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/entity/KyuubiStatementInfo.scala [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+    appId: String,
+    sessionId: HandleIdentifier,
+    var queryExecution: QueryExecution,
+    var exception: KyuubiSQLException,
```
Comment (ulysses-you) [inline externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/entity/KyuubiStatementInfo.scala:45]: can we move these `var` inside class ?

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/KyuubiStatementMonitorSuite.scala [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+        req.setStatement(sql)
+        val tExecuteStatementResp = client.ExecuteStatement(req)
+        val operationHandle = tExecuteStatementResp.getOperationHandle
+        waitForOperationToComplete(client, operationHandle)
+      }
+
+      var iterator = kyuubiStatementQueue.iterator()
+      while (iterator.hasNext) {
+        val kyuubiStatementInfo = iterator.next()
+        assert(kyuubiStatementInfo.statement !== null)
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/KyuubiStatementMonitorSuite.scala:58]: can we check the real value directly ? `!= null` seems unreliable

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/KyuubiStatementMonitorSuite.scala [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+  override def withKyuubiConf: Map[String, String] = Map.empty
+
+  test("add kyuubiStatementInfo into queue and remove them by size type threshold") {
+    val sql = "select timestamp'2021-06-01'"
+    val total: Int = 7
+    // Clear kyuubiStatementQueue first
+    val getQueue = PrivateMethod[
+      ArrayBlockingQueue[KyuubiStatementInfo]](Symbol("kyuubiStatementQueue"))()
+    val kyuubiStatementQueue = KyuubiStatementMonitor.invokePrivate(getQueue)
+    kyuubiStatementQueue.clear()
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/KyuubiStatementMonitorSuite.scala:44]: is it better to add a new method `reset` in `KyuubiStatementMonitor` ?

Comment (zhang1002) [reply externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/KyuubiStatementMonitorSuite.scala:44]: This method is just for unit test.
In pro_env, we don't need this method.

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/KyuubiStatementMonitorSuite.scala (hunk 24 above)
Comment (zhang1002) [reply externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/KyuubiStatementMonitorSuite.scala:58]: ok, i will do it

File: externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/entity/KyuubiStatementInfo.scala (hunk 23 above)
Comment (zhang1002) [reply externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/entity/KyuubiStatementInfo.scala:45]: ok

File: externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/KyuubiStatementMonitor.scala (hunk 20 above)
Comment (zhang1002) [reply externals/kyuubi-spark-monitor/src/main/scala/org/apache/kyuubi/engine/spark/monitor/KyuubiStatementMonitor.scala:49]: It just for test. We will make those configurable soon.

**Review by yaooqinn — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks, merging to master for v1.3.0

**Pull request merged**